### PR TITLE
Story4

### PR DESCRIPTION
**Pull request by joseph-padfield**

### CODE REVIEW

File: product.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?php
+require_once ('vendor/autoload.php');
+use FurnitureStoreApi\Services\HeaderService;
+use FurnitureStoreApi\Products\ProductHydrator;
+use FurnitureStoreApi\Services\ResponseService;
+use FurnitureStoreApi\Exceptions\InvalidProductIdException;
+HeaderService::setHeaders();
+
+
```
Comment (joseph-padfield) [inline product.php]: empty line

File: product.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    $jsonString = [];
+    if ($_GET['id']>ProductHydrator::getMaxProducts()
+    )
+    {
+        throw new InvalidProductIdException();
+    }
+    else
+    {
+        $jsonString = ProductHydrator::getProductById($_GET['id']);
+        ResponseService::makeResponse("Successfully retrieved product",$jsonString,200);
```
Comment (joseph-padfield) [inline product.php]: change $jsonstring - $productArray?

File: src/Exceptions/InvalidProductIdException.php [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<?php
+
```
Comment (joseph-padfield) [inline src/Exceptions/InvalidProductIdException.php]: get rid of this line to be consistent with other files? probs not necessary!

**Review by joseph-padfield — COMMENTED** [review]
See comments

File: categories.php [hunk 4]
```diff
@@ -7,11 +7,11 @@
 
 try
 {
-    $jsonString = CategoryHydrator::getCategories();
-    ResponseService::makeResponse('Successfully retrieved products', $jsonString,200);
+    $jsonArray = CategoryHydrator::getCategories();
+    ResponseService::makeResponse('Successfully retrieved products', $jsonArray,200);
```
Comment (matt-ioa) [inline categories.php]: It's still not technically json yet, so just call this `resultsArray` instead

File: categories.php [hunk 5]
```diff
@@ -7,11 +7,11 @@
...
 }
 catch(Exception $exception)
 {
-    $jsonString = [];
-    ResponseService::makeResponse("Unexpected Error",$jsonString, 500);
+    $jsonArray = [];
+    ResponseService::makeResponse("Unexpected Error",$jsonArray, 500);
```
Comment (matt-ioa) [inline categories.php]: No need to reassign the variable, just directly pass an empty array to the response service

File: product.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    }
+    else
+    {
+        $jsonArray = ProductHydrator::getProductById($_GET['id']);
+        ResponseService::makeResponse("Successfully retrieved product",$jsonArray,200);
+    }
+}
+catch (InvalidProductIdException $exception)
+{
+    ResponseService::makeResponse($exception-> getMessage(),$jsonArray,400);
```
Comment (matt-ioa) [inline product.php]: Again, just put an empty array here as there's a chance that the `$jsonArray` variable could have been assigned before an exception occurs.

File: products.php [hunk 7]
```diff
@@ -10,22 +10,22 @@
...
 }
 catch (InvalidCategoryException $exception)
 {
-    ResponseService::makeResponse($exception->getMessage(), $jsonString, 400);
+    ResponseService::makeResponse($exception->getMessage(), $jsonArray, 400);
 }
 catch(Exception $exception)
     {
-        ResponseService::makeResponse('Unexpected Error', $jsonString, 500);
+        ResponseService::makeResponse('Unexpected Error', $jsonArray, 500);
```
Comment (matt-ioa) [inline products.php]: See above

File: products.php [hunk 8]
```diff
@@ -10,22 +10,22 @@
 
 try
 {
-    $jsonString = [];
+    $jsonArray = [];
```
Comment (matt-ioa) [inline products.php]: No need to do this if you're passing the empty array into the error response below (as described above)

File: src/Products/DetailedProductEntity.php [hunk 9]
```diff
@@ -4,9 +4,49 @@
...
-    private int $category_id;
-    private int $width;
-    private int $height;
-    private int $depth;
-    private int $related;
+    protected int $category_id;
+    protected int $width;
+    protected int $height;
+    protected int $depth;
+    protected int $related;
```
Comment (matt-ioa) [inline src/Products/DetailedProductEntity.php]: These should just be private as we have no intention of inheriting from this class

File: src/Products/ProductHydrator.php [hunk 10]
```diff
@@ -13,4 +14,22 @@ public static function getProductsByCategory($categoryID)
         $query->setFetchMode(PDO::FETCH_CLASS, SimpleProductEntity::class);
         return $query->fetchALL();
     }
+
+    public static function getProductById($productId)
+    {
+        $db = DbConnection::setConnection();
+        $query = $db -> prepare ('SELECT `id`,`name`,`category_id`,`width`,`height`,`depth`,`price`,`stock`,`related`,`color` FROM `products` WHERE `id`= :productId');
```
Comment (matt-ioa) [inline src/Products/ProductHydrator.php]: Remove the spaces so it looks like this: `$query = $db->prepare(`

File: src/Services/ResponseService.php [hunk 11]
```diff
@@ -3,10 +3,10 @@
 namespace FurnitureStoreApi\Services;
 class ResponseService
 {
-    public static function makeResponse($message, $jsonString, $responseCode)
+    public static function makeResponse($message, $jsonArray, $responseCode)
```
Comment (matt-ioa) [inline src/Services/ResponseService.php]: Again, call this `resultsArray` everywhere as it isn't json yet

File: src/Products/ProductHydrator.php [hunk 12]
```diff
@@ -13,4 +14,22 @@ public static function getProductsByCategory($categoryID)
...
+        $query->execute(['productId'=>$productId]);
+        $query->setFetchMode(PDO::FETCH_CLASS, DetailedProductEntity::class);
+        return $query->fetch();
+    }
+    public static function getMaxProducts()
+    {
+        $db = DbConnection::setConnection();
```
Comment (matt-ioa) [inline src/Products/ProductHydrator.php]: Rename the method in `DbConnection` to `getConnection`, as it's getting an existing connection

Comment (matt-ioa) [reply src/Products/ProductHydrator.php]: (even though it sets it the first time!)

**Review by matt-ioa — CHANGES_REQUESTED** [review]
Some changes suggested, take a look 👀

**Review by matt-ioa — APPROVED** [review]

**Pull request merged**